### PR TITLE
More intuitive same-line placement for list-like values

### PR DESCRIPTION
**Pull request by FuegoFro**
This diff tries to add more intuitive/reasonable placement for values in certain situations. In particular, if we have list-like literals that are too long to fit on a line, we often want the open/close brackets of the literal to be on the same line as the surrounding structure, rather than be entirely on their own line. An example is:

```
foo([
    1,
    2,
    ...
    10,
])
```

The naive way of using `group`s would cause the `[` and `]` to each be on their own lines, due to how `group` splits from the outside in. This adds optimizations/special cases for this style of formatting in `Assign`s (and the closely related `AnnAssign`s and `AugAssign`s), `Call`s, `Return`s and `Yield`s.

There are fairly extensive tests for the new behavior, but let me know if there's any other cases you'd like to see! In particular there's tests for non-list-like expressions to make sure that "normal" wrapping works as indended, and tests for the various kinds of assignments. There's still some work for how we handle complex (multi-part) strings, but I think that's mostly unrelated to this diff.

I also tried to comment it so it'd be easier to read through in the future, but let me know if there's any part that is confusing or could have better comments!

### CODE REVIEW

**Comment by azz** [comment]
I've been super busy the last couple of weeks, sorry! I'll try and find some time to review this properly soon, but I'd just offer a word of caution with `conditionalGroup`: it can be quite computationally expensive as it basically causes exponential branches in the doc printer.

**Comment by FuegoFro** [comment]
No rush!

Regarding `conditionalGroup`, I'd love any suggestions for simpler alternatives of how to implement this behavior! I got the ideas for this from looking at the [implementation of function call formatting](https://github.com/prettier/prettier/blob/master/src/language-js/printer-estree.js#L3196) in the js formatter in the main Prettier repo. From my understanding of how `group` works, it doesn't seem possible to get this behavior any other way, but I'd love to hear any ideas you have!

And regarding the exponential time complexity, my understanding from reading `doc-printer.js` is that this has basically the same properties as normal `group`. In particular, my understanding of [the algorithm](https://github.com/prettier/prettier/blob/master/src/doc/doc-printer.js#L241) is:

- See if the first conditional state fits flat
- If not, try the next conditional state, again flat
- If we get to the last conditional state and it doesn't fit flat, then try breaking that one.

Additionally, it doesn't look like [the `fits` function](https://github.com/prettier/prettier/blob/master/src/doc/doc-printer.js#L112) does any backtracking or breaking of groups (or looking at subsequent expanded states of groups).

Given this, I don't think nested `conditionalGroups` actually are exponential? But I could definitely be wrong!

**Comment by FuegoFro** [comment]
@azz if you have some time, it'd be awesome if you could take a glance over some of this! I think the most relevant parts to look at would be the `printExpandingValue` and `printMultiAssignment` functions :)

**Comment by azz** [comment]
Also r.e. complexity, I have never actually done the analysis on the code, I think it was @vjeux who mentioned that it was exponential? I seem to recall performance issues in Prettier JS due to it but I could be mistaken.

File: tests/python_keep_opening_on_previous_line/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+]
+
+my_long_variable_1_aaaaaaaaaaaaaaaaaaaaaa = \\
+    my_long_variable_2_aaaaaaaaaaaaaaaaaaaaaa = [
+        long_item_1_aaaaaaaaaaaa,
+        long_item_1_aaaaaaaaaaaa,
+        long_item_1_aaaaaaaaaaaa
+    ]
+
+foo(bar(baz([
```
Comment (vjeux) [inline tests/python_keep_opening_on_previous_line/__snapshots__/jsfmt.spec.js.snap]: The way prettier algorithm has been designed is to break the outermost group first, so this should look the following way:

```js
foo(
  bar(
    baz(
      ...
    )
  )
)
```

If you want it to print it the way you are displaying it right now, prettier is --not-- the right algorithm. And frankly, I don't know what the right algorithm is...

File: tests/python_keep_opening_on_previous_line/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    ],
+    1
+)))
+
+foo(bar(
+    baz([
+        long_item_1_aaaaaaaaaaaa,
+        long_item_1_aaaaaaaaaaaa,
+        long_item_1_aaaaaaaaaaaa
+    ]),
```
Comment (vjeux) [inline tests/python_keep_opening_on_previous_line/__snapshots__/jsfmt.spec.js.snap]: It seems like your goal is to not have `[]` be on its own line. You don't need conditionalGroup for this. What you can do when you are printing a function call is to look at the type of the arguments and if there's a single argument which is an array, then output something different. The whole game of prettier is from the ast to figure out what IR to output such that you get the right thing at the end.

File: tests/python_keep_opening_on_previous_line/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    1 / my_long_value_aaaaaaaaaaaaaaaaaaaaaaaaaa
+
+my_long_variable_1_aaaaaaaaaaaaaaaaaaaaaa = \\
+    my_long_variable_2_aaaaaaaaaaaaaaaaaaaaaa = \\
+    my_long_value_aaaaaaaaaaaaaaaaaaaaaaaaaa
+
+my_long_variable_1_aaaaaaaaaaaaaaaaaaaaaa = \\
+    my_long_variable_2_aaaaaaaaaaaaaaaaaaaaaa = \\
+    my_long_variable_3_aaaaaaaaaaaaaaaaaaaaaa = \\
+    my_long_value_aaaaaaaaaaaaaaaaaaaaaaaaaa
```
Comment (vjeux) [inline tests/python_keep_opening_on_previous_line/__snapshots__/jsfmt.spec.js.snap]: You should be able to get the result that you want without conditional groups using carefully crafted groups, indents and ifBreak.

File: tests/python_keep_opening_on_previous_line/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (azz) [reply tests/python_keep_opening_on_previous_line/__snapshots__/jsfmt.spec.js.snap]: Yeah that's a good point. How crucial is the inlineless of `foo(bar(baz(` to Python style?

**Comment by FuegoFro** [comment]
Sorry for leaving this open for so long! I haven't had time to work on this in a while, so I'm going to go ahead and close this out. If people still feel this is valuable, we can take another stab at addressing some of the complexity concerns.

For what it's worth, I still am not convinced this is actually any slower or more asymptotically complex than normal `group`, but again that discussion could happen at a later time.

**Pull request closed without merging**